src/engine/strat_engine/thinpool/thinpool.rs

```diff
@@ -351,45 +362,47 @@ impl ThinPool {
         Ok(extend_size)
     }
 
-    /// Extend the thinpool with new data regions.
-    fn extend_data(&mut self, dm: &DM, new_segs: &[BlkDevSegment]) -> EngineResult<()> {
-        let mut segments = Vec::with_capacity(self.data_segments.len() + new_segs.len());
-        segments.extend_from_slice(&self.data_segments);
-
-        // Last existing and first new may be contiguous. Coalesce into
-        // a single BlkDevSegment if so.
-        let coalesced_new_first = {
-            match new_segs.first() {
-                Some(new_first) => {
-                    let old_last = segments
-                        .last_mut()
-                        .expect("thin pool must always have some data segments");
-                    if old_last.uuid == new_first.uuid &&
-                       (old_last.segment.start + old_last.segment.length ==
-                        new_first.segment.start) {
-                        old_last.segment.length += new_first.segment.length;
-                        true
-                    } else {
-                        false
-                    }
-                }
-                None => false,
-            }
-        };
-
-        if coalesced_new_first {
-            segments.extend_from_slice(&new_segs[1..]);
+    /// Expand the physical space allocated to a pool meta by extend_size.
+    /// Return the number of MetaBlocks added.
+    fn extend_thinpool_meta(&mut self,
+                            dm: &DM,
+                            extend_size: MetaBlocks,
+                            bd_mgr: &mut BlockDevMgr)
+                            -> EngineResult<MetaBlocks> {
+        if let Some(mut new_meta_regions) = bd_mgr.alloc_space(&[extend_size.sectors()]) {
+            self.extend_meta(dm,
+                             &new_meta_regions
+                                  .pop()
```

Comment:
I doubt you need to pop here, just use ```first()```.   ```new_meta_regions``` can then be immutable. Might as well fix up ```extend_thinpool``` which does same thing in separate commit.

Reply:
ok